### PR TITLE
refactor: correct types for AbstractOperation.execute

### PR DESCRIPTION
**Pull request by nbbeeken**
#### What is changing?

The session can sometimes be undefined in execOperation, we luckily didn't have any code that relied on its existence :)

* made explicit use of override
* made explicit use of return, caught bug in withSession
* Updated eslint typescript plugin so the parser stops mentioning TS version mismatch

#### What is the motivation for this change?

Correct internal types, should have no impact on users, helps devs reason about code.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/bulk/common.ts [hunk 1]
```diff
@@ -289,6 +289,7 @@ export class BulkWriteResult {
     if (index < this.result.writeErrors.length) {
       return this.result.writeErrors[index];
     }
+    return;
```
Comment (baileympearson) [inline src/bulk/common.ts:292]:
```suggestion
return index < this.result.writeErrors.length ? this.result.writeErrors[index] : undefined
```

Comment (nbbeeken) [reply src/bulk/common.ts:292]: 🧑‍💻  ⛳ ?
I'm okay with the suggestion but I don't think we need to be writing `return undefined;` anywhere, which I know this isn't but it's implicitly similar. Does that make sense?

File: package.json [hunk 2]
```diff
@@ -52,15 +52,15 @@
     "@types/semver": "^7.3.9",
     "@types/sinon": "^10.0.11",
     "@types/whatwg-url": "^8.2.1",
-    "@typescript-eslint/eslint-plugin": "^5.13.0",
-    "@typescript-eslint/parser": "^5.13.0",
+    "@typescript-eslint/eslint-plugin": "^5.14.0",
```
Comment (baileympearson) [inline package.json:55]: sneaking in some version bumps I see, classic classic

File: src/mongo_client.ts [hunk 3]
```diff
@@ -556,52 +549,34 @@ export class MongoClient extends TypedEventEmitter<MongoClientEvents> {
    * NOTE: presently the operation MUST return a Promise (either explicit or implicitly as an async function)
    *
    * @param options - Optional settings for the command
-   * @param callback - An callback to execute with an implicitly created session
+   * @param withSessionCallback - An callback to execute with an implicitly created session
```
Comment (baileympearson) [inline src/mongo_client.ts:552]: The name of the parameter is still callback.  Let's either rename the parameter or revert this change

File: src/bulk/common.ts (hunk 1 above)
Comment (nbbeeken) [reply src/bulk/common.ts:292]: What about `this.result.writeErrors[index]`?

File: src/error.ts [hunk 4]
```diff
@@ -670,7 +670,6 @@ function makeWriteConcernResultObject(input: any) {
 export class MongoWriteConcernError extends MongoServerError {
   /** The result document (provided if ok: 1) */
   result?: Document;
```
Comment (baileympearson) [inline src/error.ts:672]: Did you mean to remove `errInfo` from the class?

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: src/mongo_client.ts (hunk 3 above)
Comment (nbbeeken) [reply src/mongo_client.ts:552]: Good catch I say we leave it as is

File: src/bulk/common.ts (hunk 1 above)
Comment (baileympearson) [reply src/bulk/common.ts]: For some reason I thought that threw an error if index was out of range.  I guess not - I'd prefer this.

I actually prefer `return undefined` but that's just preference.  I like that it's explicit 🤷 but I don't feel terribly strongly about this

File: src/error.ts (hunk 4 above)
Comment (nbbeeken) [reply src/error.ts:672]: Yes it's already defined in the parent, you'll notice a few other removals, if the override doesn't actually narrow the type then there's no purpose for it.

File: src/bulk/common.ts (hunk 1 above)
Comment (dariakp) [reply src/bulk/common.ts]: return undefined would read weird to me personally, and in general, and especially in holding with TS conception of types, if your function is supposed to return something but you don't have a result, the appropriate default is `null` not undefined

File: src/connection_string.ts [hunk 5]
```diff
@@ -955,16 +957,14 @@ export const OPTIONS = {
           readPreference: { ...options.readPreference, ...value },
           ...value
         } as any);
-      }
-      if (isRecord(value, ['mode'] as const)) {
+      } else if (isRecord(value, ['mode'] as const)) {
```
Comment (dariakp) [inline src/connection_string.ts:960]: why the additions of `else` here all the returns are already definitely typed, so this seems like unnecessary logic, minus the last throw; if anything, this syntax implies that the block above might *not* properly return

File: src/cursor/find_cursor.ts [hunk 6]
```diff
@@ -329,21 +329,6 @@ export class FindCursor<TSchema = Document> extends AbstractCursor<TSchema> {
     return this;
   }
 
-  /**
-   * Set a maxTimeMS on the cursor query, allowing for hard timeout limits on queries (Only supported on MongoDB 2.6 or higher)
-   *
-   * @param value - Number of milliseconds to wait before aborting the query.
-   */
-  maxTimeMS(value: number): this {
```
Comment (dariakp) [inline src/cursor/find_cursor.ts:337]: does it matter that this used to be defined on `kBuiltOptions`, which gets spread during initialize on L73, and now it's deeper on the `cursorOptions`, which second in the spread? the comment there explicitly says that the order matters, i.e. that those cursorOptions take priority over the find cursor options, and, while that seems backwards, it's a little bit ominous, so I just wanted to know if you explicitly looked into that or have any context on the comment

File: src/bulk/common.ts (hunk 1 above)
Comment (nbbeeken) [reply src/bulk/common.ts]: I agree, I went with the ternary though to preserve the API. Filed NODE-4068

File: src/mongo_client.ts [hunk 7]
```diff
@@ -564,44 +557,26 @@ export class MongoClient extends TypedEventEmitter<MongoClientEvents> {
...
-      const result = callback(session);
-      return Promise.resolve(result).then(
-        result => cleanupHandler(undefined, result, undefined),
-        err => cleanupHandler(err, null, { throw: true })
-      );
-    } catch (err) {
-      return cleanupHandler(err, null, { throw: false }) as Promise<void>;
-    }
+    return Promise.resolve()
+      .then(() => withSessionCallback(session))
```
Comment (dariakp) [inline src/mongo_client.ts:578]: technically, because the result is returned, the declared return type is incorrect, because the type returned would be the type of the return type of `withSessionCallback` and not guaranteed to be `Promise<void>`

**Review by dariakp — CHANGES_REQUESTED** [review]
Some questions comments

File: src/connection_string.ts (hunk 5 above)
Comment (nbbeeken) [reply src/connection_string.ts]: Good point, there was no need for else(s), reverted, kept the throws.

File: src/cursor/find_cursor.ts (hunk 6 above)
Comment (nbbeeken) [reply src/cursor/find_cursor.ts:337]: Ah kBuiltOptions vs kOptions, really good catch I did not catch the difference in the super class. I've reverted

File: src/mongo_client.ts (hunk 7 above)
Comment (nbbeeken) [reply src/mongo_client.ts]: fixed!

File: test/integration/sessions/sessions.test.js [hunk 8]
```diff
@@ -72,84 +72,59 @@ describe('Sessions Spec', function () {
...
-
-          return promise
-            .then(() => expect(client.topology.s.sessionPool.sessions).to.have.length(1))
+          return client
+            .withSession(testCase.operation(client))
+            .then(
+              () => expect(client.topology.s.sessionPool.sessions).to.have.length(1),
+              () => expect(client.topology.s.sessionPool.sessions).to.have.length(1)
+            )
             .then(() => client.close())
```
Comment (dariakp) [inline test/integration/sessions/sessions.test.js:128]: the client close here and in the test case below should probably be in an afterEach; the test cases should also differentiate `resolve` from `reject` cases in their description and corresponding assertions; the test case descriptions should indicate that we care that sessions are cleaned up in both cases

**Review by dariakp — CHANGES_REQUESTED** [review]

**Review by dariakp — DISMISSED** [review]
LGTM!

File: src/connection_string.ts [hunk 9]
```diff
@@ -975,6 +977,7 @@ export const OPTIONS = {
           rpOpts
         );
       }
+      throw new MongoParseError(`Unknown ReadPreference value: ${value}`);
```
Comment (baileympearson) [inline src/connection_string.ts:980]: Could we add a test for this new error message if it's not covered in our spec tests already?

File: test/integration/sessions/sessions.test.ts [hunk 10]
```diff
@@ -72,93 +73,108 @@ describe('Sessions Spec', function () {
...
-          },
-          {
-            description: 'should support operations that throw exceptions',
-            operation: (/* client */) => (/* session */) => {
-              throw new Error('something went wrong!');
-            }
+      const tests = [
+        {
+          description: 'should resolve non-async callbacks that return promises',
+          operation: client => session => {
```
Comment (baileympearson) [inline test/integration/sessions/sessions.test.ts:89]: Is there any reason we couldn't have a single function with two parameters?  We're not currying anywhere so I don't think this is necessary

File: test/integration/sessions/sessions.test.ts [hunk 11]
```diff
@@ -72,93 +73,108 @@ describe('Sessions Spec', function () {
...
+            throw new Error('something went wrong!');
+          }
+        }
+      ];
 
-        it('supports passing options to ClientSession', function () {
-          const client = test.client;
+      for (const testCase of tests) {
+        it(testCase.description, async function () {
+          const shouldResolve = testCase.description.startsWith('should resolve');
```
Comment (baileympearson) [inline test/integration/sessions/sessions.test.ts:129]: This isn't blocking - but could we just add a property to the tests instead to indicate whether or not the test should resolve or reject?

```typescript
{
      description: 'should reject callbacks that throw synchronous exceptions',
      operation: (/* client */) => (/* session */) => {
        throw new Error('something went wrong!');
      },
      result: 'resolve' // or something like this
}

...

const shoudResolve = result === 'resolve';
```

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: test/integration/sessions/sessions.test.ts (hunk 10 above)
Comment (nbbeeken) [reply test/integration/sessions/sessions.test.ts:89]: The function returned from the first call has to be correctly shaped for withSession, withSession only passes in the session as a single argument. That being said after updating this test to the before/after style the client is accessible from the outer scope so we don't need the factory style anymore

File: src/connection_string.ts (hunk 9 above)
Comment (nbbeeken) [reply src/connection_string.ts:980]: Yes should be a quick unit test, good point

**Review by dariakp — APPROVED** [review]
